### PR TITLE
APPS/`pkeyutl`: improve `-rawin` usability and doc for EdDSA

### PR DESCRIPTION
**Pull request by DDvO**
This improves the usability of they `pkeyutl` app when signing or verifying with Ed25519 and Ed448 keys.

When the `-rawin` option was not given with them, so far an obscure low-level error was thrown such as
```
pkeyutl: Error initializing context
80205AEC01000000:error:03000096:digital envelope routines:evp_pkey_signature_init:operation not supported for this keytype:crypto/evp/signature.c:526:
```
This PR makes sure that the `-rawin` option is implied for such scenarios, 
[**update**: also in case the `-digest` option is given, ]
also making the app more convenient to use.
Implementing this required a split of the `init_ctx()` helper function, which is now less convoluted.

Tests and doc are updated accordingly,
and on this occasion the help output and the doc of the `-rawin` option are made less confusing.

##### Checklist

- [x] documentation is added or updated
- [x] tests are added or updated

### CODE REVIEW

**Review by t-j-h — APPROVED** [review]

File: apps/pkeyutl.c
Comment (hlandau) [inline apps/pkeyutl.c:305]: How is pkey freed in this case?

**Comment by beldmit** [comment]
@DDvO could you please answer @hlandau's comment?
Willit work with ed448/25519 when `-rawin` is specified (scripts backward compatibility)?

File: apps/pkeyutl.c
Comment (DDvO) [reply apps/pkeyutl.c:305]: `pkey` gets freed as in the usual way, 
as can be seen easily following the `opthelp` and then `end` labels.

Comment (DDvO) [reply apps/pkeyutl.c:305]: Oops, I had some confusion; will fix the mem leak...

**Comment by DDvO** [comment]
> @DDvO could you please answer @hlandau's comment?

Thanks for pointing me to that, and
sorry that I missed that good review comment of March 10th.
Handled it now.

> Will it work with ed448/25519 when `-rawin` is specified (scripts backward compatibility)?

Yes.

Rebased to fix trivial merge conflict in `CHANGES.md`.

File: apps/pkeyutl.c
Comment (beldmit) [inline apps/pkeyutl.c:52]: Shouldn't it be `_is_a` instead ?

File: test/recipes/20-test_pkeyutl.t
Comment (beldmit) [inline test/recipes/20-test_pkeyutl.t:81]: Isn't it worth adding a test _with_ `-rawin` for backward compatibility testing?

**Review by beldmit — COMMENTED** [review]
Two minor questions

File: apps/pkeyutl.c
Comment (DDvO) [reply apps/pkeyutl.c:52]: Can do - which will be the right string argument?

File: test/recipes/20-test_pkeyutl.t
Comment (DDvO) [reply test/recipes/20-test_pkeyutl.t:81]: Ok, will do...

File: apps/pkeyutl.c
Comment (DDvO) [reply apps/pkeyutl.c:52]: Hmm, I found one cannot use `EDDSA` or the like.
Would this be any better?
```
    return EVP_PKEY_is_a(pkey, "ED25519")
        || EVP_PKEY_is_a(pkey, "ED448");
```
Likely it's less efficient.

Comment (beldmit) [reply apps/pkeyutl.c:52]: I wouldn't worry about effectiveness here as the command-line util is not intended to be very effective

Comment (DDvO) [reply apps/pkeyutl.c:52]: Right, efficiency does not really matter here.
So changed it this way.

File: test/recipes/20-test_pkeyutl.t
Comment (DDvO) [reply test/recipes/20-test_pkeyutl.t]: Done.

File: apps/pkeyutl.c
Comment (beldmit) [reply apps/pkeyutl.c:52]: Thanks!

**Review by beldmit — APPROVED** [review]
LGTM

File: test/recipes/20-test_pkeyutl.t
Comment (beldmit) [reply test/recipes/20-test_pkeyutl.t]: Thanks!

**Comment by beldmit** [comment]
@t-j-h would you mind confirming your aproval?

**Comment by DDvO** [comment]
@hlandau or can you give the 2nd approval?

**Comment by DDvO** [comment]
Rebased to fix trivial merge conflict

**Comment by DDvO** [comment]
Rebased again to fix trivial merge conflict in `CHANGES.md`

**Comment by DDvO** [comment]
PIng @openssl/committers for 2nd approval / reconfirmation.
Should not be a big deal.

File: doc/man1/openssl-pkeyutl.pod.in
Comment (DDvO) [reply doc/man1/openssl-pkeyutl.pod.in:69]: Good point,
though support for the prehash variant (HashEdDSA) seems unlikely (see #22910 23441).

Yes, the `-rawin` option would still be  implied in this case,
and I've added a check whether the `-digest` optionh been given in addition, with an error message pointing out that the ph variant is not supported.

**Comment by DDvO** [comment]
I also have to ask also @t-j-h to reconfirm approval.

**Review by beldmit — APPROVED** [review]
LGTM

**Comment by DDvO** [comment]
Asking again @t-j-h to reconfirm, or any @openssl/committers for 2nd review.

File: apps/pkeyutl.c
Comment (vdukhovni) [inline apps/pkeyutl.c:51]: Should anything be done to ensure that ECDSA signatures are with a digest of ideally the same (or perhaps shorter) length, since the underlying ECDSA signature scheme truncates inputs.

Indeed, one would hope the ECDSA signatures are always with the *default* digest for the algorithm, making `-rawin` the preferred way for pretty much all operations, feeding in already digested data is most unusual, and requires care to get it right.

File: apps/pkeyutl.c
Comment (vdukhovni) [inline apps/pkeyutl.c:345]: This is supposed to **take ownership** of `pkey`, but it can return `NULL` early on error, without freeing `pkey`, leaving it dangling.

File: apps/pkeyutl.c
Comment (vdukhovni) [inline apps/pkeyutl.c:648]: This fails to free `pkey`.

File: apps/pkeyutl.c
Comment (vdukhovni) [inline apps/pkeyutl.c:668]: This fails to free `pkey` (in the KDF code path).

Comment (vdukhovni) [inline apps/pkeyutl.c:668]: This fails to free `pkey` (in the KDF code path).

**Review by vdukhovni — CHANGES_REQUESTED** [review]

File: apps/pkeyutl.c
Comment (beldmit) [inline apps/pkeyutl.c:29]: I don't like this change...

File: CHANGES.md
Comment (DDvO) [reply CHANGES.md]: Oops, the plural was certainly wrong.
Changed as suggested.

Doing so, I found that `CHANGES.md` includes some more mentions of the words 'app' and 'apps' - 
for consistency, changed to 'command' and 'commands', respectively (In a separate commit).

File: apps/pkeyutl.c
Comment (DDvO) [reply apps/pkeyutl.c:343]: Oh, good spotting!
I've now fixed all those potential memore leak cases in `init_ctx()`.

File: apps/pkeyutl.c
Comment (DDvO) [reply apps/pkeyutl.c]: Oops, I don't like it either ;-)
I forgot to copy over to here the comment I had already changed  in the function definition header; fixed now.

File: apps/pkeyutl.c
Comment (DDvO) [reply apps/pkeyutl.c:51]: Very good point!!
I found that the documentation falsely claims on the `-digest` option:

> If this option is omitted but the signature algorithm requires one, a default value will be
> used. For signature algorithms like RSA, DSA and ECDSA, SHA-256 will be the default digest algorithm.
> For SM2, it will be SM3.

I'm going to file a bug report on that later today...

For now, I've just strengthened the existing error message in case the input is more than 512 bits long
without a digest given, which is a clear indication that the user failed to provide it.

I'd like to add a warning (to allow for backward compatibility and for special cases where this may be intended)
in case the input is longer than the input size of the asymmetric signing operation.
To this end: is there an API function for getting that value for any given EVP_PKEY?

Comment (vdukhovni) [reply apps/pkeyutl.c:51]: > Very good point!! I found that the documentation falsely claims on the `-digest` option:

The claim does not look false to me, whether signing and verification work equally well with an implicit digest.
```
$ openssl pkeyutl -in msg -sign -inkey p256.pem -rawin -out p256-sig-default.dat
$ openssl pkeyutl -in msg -sign -inkey p256.pem -rawin -digest sha256 -out p256-sig-sha256.dat

$ openssl pkeyutl -rawin -in msg -sigfile p256-sig-default.dat -verify -inkey p256.pem
Signature Verified Successfully

$ openssl pkeyutl -rawin -in msg -sigfile p256-sig-default.dat -digest sha256 -verify -inkey p256.pem
Signature Verified Successfully

$ openssl pkeyutl -rawin -in msg -sigfile p256-sig-sha256.dat -digest sha256 -verify -inkey p256.pem
Signature Verified Successfully

$ openssl pkeyutl -rawin -in msg -sigfile p256-sig-sha256.dat -verify -inkey p256.pem
Signature Verified Successfully
```

File: apps/pkeyutl.c
Comment (vdukhovni) [inline apps/pkeyutl.c:664]: Set `pkey =  NULL` here to avoid double-free?

File: apps/pkeyutl.c
Comment (vdukhovni) [inline apps/pkeyutl.c:293]: This is wrong, because when the input **is** a digested value, one might (especially with RSA PKCS#1) want to indicate which digest the input was constructed with. The `-digest <...>` option is in fact especially useful with `rawin` false, with `rawin` true, the default digest is generally a sensible choice, though with RSA there are often multiple valid choices (RSA with SHA2-256 and RSA with SHA2-512).

**Review by vdukhovni — CHANGES_REQUESTED** [review]

File: apps/pkeyutl.c
Comment (DDvO) [reply apps/pkeyutl.c:51]: You did not say which size `msg` has in your case,
but likely this is immaterial for your above tests because you use `-rawin` in all of them.
Yet the bug, as reported in #25827, is on the case that `-rawin` is not given.

Note that the above quoted documentation of `-digest` does not include the precondition that `-rawin` is given.
Maybe the issue is simply that this was forgotten to mention?

Comment (levitte) [reply apps/pkeyutl.c:51]: Actually, the `-digest` document does say that it's only valid together with `-rawin`.  Quoting just part of the `-digest` documentation hides this detail.

File: apps/pkeyutl.c
Comment (levitte) [reply apps/pkeyutl.c:293]: You're mixing up `-digest {name}` with `-pkeyopt digest:{name}`.  The latter is used when the input is a digested value (and it usually has a default too).

File: apps/pkeyutl.c
Comment (vdukhovni) [reply apps/pkeyutl.c:51]: So how exactly is one supposed to sign pre-digested input with RSA PKCS#1? Without `-digest` it seems that the `rawin` input needs to include the PKCS#1 digest OID and parameters, as well as ASN.1 encoding of the digest, otherwise I get:

```
$ openssl pkeyutl -in msghash.dat -sign -inkey rsa.pem -out rsa-sig-default.dat

$ openssl pkeyutl -inkey rsa.pem -encrypt -pkeyopt rsa_padding_mode:none -in rsa-sig-default.dat | od -tx1
0000000 00 01 ff ff ff ff ff ff ff ff ff ff ff ff ff ff
0000020 ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff
*
0000360 00 41 74 74 61 63 6b 20 61 74 20 64 61 77 6e 0a
```
In which the PKCS#1 payload is just the hash, which isn't that useful.

Comment (levitte) [inline apps/pkeyutl.c:51]: This, BTW, just isn't right.  The categorization should be made on capability; in this case, which API the implementation supports.  capability can be detected by trying to initialize the operation with `EVP_Digest{Whatever}Init()` and `EVP_PKEY_{whatever}_init()` (and `EVP_PKEY_{whatever}_message_init()` when we've landed that support)

Comment (DDvO) [reply apps/pkeyutl.c:51]: > Actually, the `-digest` document does say that it's only valid together with `-rawin`. 
> Quoting just part of the `-digest` documentation hides this detail.

Not really - the doc says there:
> If this option is present, then the `-rawin` option must be also specified.

So -digest cannot be used without -rawin, but when -digest is omitted, nothing is implied for -rawin.
The -rawin option is pretty confusing, and the aim of this PR to make it more or less implicit.

Comment (DDvO) [reply apps/pkeyutl.c:51]: The `-rawin` option is pretty confusing, and the aim of this PR is, for usability. to make it more or less implicit.

Comment (vdukhovni) [reply apps/pkeyutl.c:51]: Though, admittedly in the case of ECDSA (IIRC) the digest OID is not part of the signed data, one just signs the digest value, and it is up to verifyer to know the digest in advance. So the need to encode a digest OID in the signature input looks RSA-specific.

Comment (DDvO) [reply apps/pkeyutl.c:51]: Yeah, I had asked about a smarter and more to-the-point way of doing this check.
Would be good if there was a general and simple-to-use EVP_PKEY API call for it.

Comment (levitte) [reply apps/pkeyutl.c:51]: Actually, what you're doing there is documented in the NOTES about `rsa_padding_mode:{mode}`:

https://github.com/openssl/openssl/blob/a3660729e68dc11c01edb4a349ff2610b6b59ee0/doc/man1/openssl-pkeyutl.pod.in#L273-L276

The "secret" for setting the digest in this case is another `-pkeyopt`, see the beginning of NOTES:

https://github.com/openssl/openssl/blob/a3660729e68dc11c01edb4a349ff2610b6b59ee0/doc/man1/openssl-pkeyutl.pod.in#L238-L239

So your experiment gave the documented result, and the way to get a signature containing a full DigestInfo looks like this:

``` console
$ openssl pkeyutl -in <(echo whatever | openssl sha256 -binary) -sign -inkey rsa.pem -pkeyopt digest:sha256 -out rsa-sig-default.dat
$ openssl pkeyutl -inkey rsa.pem -encrypt -pkeyopt rsa_padding_mode:none -in rsa-sig-default.dat | od -tx1
0000000 00 01 ff ff ff ff ff ff ff ff ff ff ff ff ff ff
0000020 ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff
*
0000300 ff ff ff ff ff ff ff ff ff ff ff ff 00 30 31 30
0000320 0d 06 09 60 86 48 01 65 03 04 02 01 05 00 04 20
0000340 cd 29 3b e6 ce a0 34 bd 45 a0 35 27 75 a2 19 ef
0000360 5d c7 82 5c e5 5d 1f 7d ae 97 62 d8 0c e6 44 11
0000400
```

Comment (levitte) [reply apps/pkeyutl.c:51]: > So the need to encode a digest OID in the signature input looks RSA-specific.

Yup...  or to be more precise, it is specific for PKCS#1.  That's why you have to give it as a `-pkeyopt`, 'cause it's only applicable for some algorithms, not others.

File: CHANGES.md
Comment (DDvO) [reply CHANGES.md]: Yes - fixed my mistake.

File: apps/pkeyutl.c
Comment (DDvO) [reply apps/pkeyutl.c:664]: Indeed this is meanwhile needed (due to the  following `goto end` in case `ctx == NULL`).
Fixed.

File: apps/pkeyutl.c
Comment (levitte) [reply apps/pkeyutl.c:51]: Agreed, it would be nice if there was a consistent capability query API for this sort of thing.

File: apps/pkeyutl.c
Comment (vdukhovni) [reply apps/pkeyutl.c:293]: Thanks for the clue. Good to know.

File: CHANGES.md
Comment (vdukhovni) [inline CHANGES.md:211]:
```suggestion
 * The `-rawin` option of the `pkeyutl` command is now implied (and thus no longer
```

File: apps/pkeyutl.c
Comment (vdukhovni) [inline apps/pkeyutl.c:497]:
```suggestion
                       "Error: The non-raw input data length %d is too long - max supported hashed size is %d\n",
```

File: doc/man1/openssl-pkeyutl.pod.in
Comment (vdukhovni) [inline doc/man1/openssl-pkeyutl.pod.in:75]: Should there bit a `HISTORY` or `NOTES` section that mentions the behaviour changes in this PR (mostly that `-digest` implying `-rawin` is new with OpenSSL 3.5 (or whichever is the oldest release to which this is backported).

Should there be some mention of `-pkeyopt digest:md` for RSA PKCS#1 when not `-rawin`?

File: test/recipes/20-test_pkeyutl.t
Comment (vdukhovni) [inline test/recipes/20-test_pkeyutl.t:92]: Perhaps also some tests with RSA signing of the digest value (not `-rawin`) and explicit `-pkeyopt digest:md`, then verify the raw message with `-digest`.

**Review by vdukhovni — CHANGES_REQUESTED** [review]
Pretty close now, just final polish. Thanks for all the work on this.

**Comment by levitte** [comment]
Regarding `-pkeyopt digest:{name}`, it's not just useful with RSA.  If you `openssl list -signature-algorithms -verbose`, you'll find this (assuming a default installation):

```
$ openssl list -signature-algorithms -verbose
  { 1.2.840.113549.1.1.1, 2.5.8.1.1, RSA, rsaEncryption } @ default
    settable operation parameters:
      digest: UTF8 encoded string (arbitrary size)
...
  ECDSA @ default
    settable operation parameters:
      digest: UTF8 encoded string (arbitrary size)
...
```

**Comment by vdukhovni** [comment]
> Regarding `-pkeyopt digest:{name}`, it's not just useful with RSA. If you `openssl list -signature-algorithms -verbose`, you'll find this (assuming a default installation):
> 
> ```
> $ openssl list -signature-algorithms -verbose
>   { 1.2.840.113549.1.1.1, 2.5.8.1.1, RSA, rsaEncryption } @ default
>     settable operation parameters:
>       digest: UTF8 encoded string (arbitrary size)
> ...
>   ECDSA @ default
>     settable operation parameters:
>       digest: UTF8 encoded string (arbitrary size)
> ...
> ```

What is the effect of `-pkeyopt digest:<md>` with ECDSA? It does not seem to make any difference to signature verification (and so perhaps also signature content).

File: CHANGES.md
Comment (DDvO) [reply CHANGES.md]: done

File: apps/pkeyutl.c
Comment (DDvO) [reply apps/pkeyutl.c]: done

File: doc/man1/openssl-pkeyutl.pod.in
Comment (DDvO) [reply doc/man1/openssl-pkeyutl.pod.in:75]: I've added to the HISTORY section:
```
Since OpenSSL 3.5,
the B<-digest> option implies B<-rawin>, and these two options are
no longer required when signing or verifying with an Ed25519 or Ed448 key.

```
What do you and to others think - can we backport all parts of this PR?

Comment (DDvO) [reply doc/man1/openssl-pkeyutl.pod.in:75]: > Should there be some mention of -pkeyopt digest:md for RSA PKCS#1 when not -rawin?

I'm not familiar with `-pkeyopt digest:md` and the like, so I may take some time and need to ask for more concrete input on it, in particular for being able to add suitable tests.
Should we actually consider handling `-pkeyopt digest:md` as part of this PR?

File: test/recipes/20-test_pkeyutl.t
Comment (DDvO) [reply test/recipes/20-test_pkeyutl.t:92]: I think this apparently difficult `-pkeyopt digest:md` topic should be  out of scope of this PR.

**Comment by DDvO** [comment]
Moved doc fixes that clearly should be backported to PR #25831.

**Comment by DDvO** [comment]
How about backporting the change that `-digest` implies `-rawin`?
This should be pretty convenient, so would be nice to have also in older versions.

File: doc/man1/openssl-pkeyutl.pod.in
Comment (vdukhovni) [inline doc/man1/openssl-pkeyutl.pod.in:84]:
```suggestion
At this time, HashEdDSA (the ph or "prehash" variant of EdDSA) is not supported,
```

File: doc/man1/openssl-pkeyutl.pod.in
Comment (vdukhovni) [inline doc/man1/openssl-pkeyutl.pod.in:81]:
```suggestion
is omitted (and B<-rawin> is used instead), but the signature algorithm requires one, a default value will be
```

**Review by vdukhovni — APPROVED** [review]
Please consider the merging the suggestions, and I'll reapprove.

File: doc/man1/openssl-pkeyutl.pod.in
Comment (DDvO) [reply doc/man1/openssl-pkeyutl.pod.in]: Done

File: doc/man1/openssl-pkeyutl.pod.in
Comment (DDvO) [reply doc/man1/openssl-pkeyutl.pod.in:81]: Fixing this is part of PR #25831 (also because that one certainly will be backported).

**Comment by DDvO** [comment]
> Please consider the merging the suggestions, and I'll reapprove.

Glad to hear!
Thank you for your excellent review and for doing a first approval.

**Review by vdukhovni — APPROVED** [review]

**Review by t8m — APPROVED** [review]

**Comment by DDvO** [comment]
Merged to master - thank you for the approvals.

**Comment by DDvO** [comment]
I'm glad that we've been able to complete this PR,
at least w.r.t its original aim, and actually went beyond that.

From my perspective, at least two points in this regard are still open:
* whether it's worth/suitable to backport part of these commits
* whether and how to warn more precisely if the input is not considered raw but its length is not beyond `EVP_MAX_MD_SIZE` = 512 bits but larger than the input size of the pure asymmetric signing operation (e.g., 256 bits ECDSA).

**Comment by levitte** [comment]
> From my perspective, at least the following three points in this regard are still open:

Raise a new issue for them?

**Comment by DDvO** [comment]
> > From my perspective, at least the following three points in this regard are still open:
> 
> Raise a new issue for them?

Done in #25892

**Pull request closed without merging**
